### PR TITLE
Gateio post only

### PR DESCRIPTION
**Pull request by samgermain**
2022-03-10T03:18:33.610Z
Node.js: v14.17.0
CCXT v1.75.47

# Future

```
% gateio createOrder XRP/USDT:USDT limit sell 1 0.758 '{"postOnly": true}' --verbose 
gateio.createOrder (XRP/USDT:USDT, limit, sell, 1, 0.758, [object Object])
fetch Request:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 
RequestHeaders:
 {
  'X-Gate-Channel-Id': 'ccxt',
  KEY: '',
  Timestamp: '1646882314',
  SIGN: '',
  'Content-Type': 'application/json'
} 
RequestBody:
 {"contract":"XRP_USDT","size":-1,"price":"0.758","tif":"poc"} 

handleRestResponse:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 201 Created 
ResponseHeaders:
 {
  'Access-Control-Allow-Headers': 'DNT,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type',
  'Access-Control-Allow-Methods': 'GET,POST,OPTIONS,DELETE,PUT',
  'Access-Control-Max-Age': '1728000',
  Connection: 'keep-alive',
  'Content-Type': 'application/json',
  Date: 'Thu, 10 Mar 2022 03:18:34 GMT',
  Server: 'openresty',
  'Transfer-Encoding': 'chunked',
  Vary: 'Origin'
} 
ResponseBody:
 {"id":136091264370,"contract":"XRP_USDT","mkfr":"0","tkfr":"0.0005","tif":"poc","is_reduce_only":false,"create_time":1646882314.959,"price":"0.758","size":-1,"refr":"0.2","left":-1,"text":"api","fill_price":"0","user":6693577,"status":"open","is_liq":false,"refu":2436035,"is_close":false,"iceberg":0}
 

2022-03-10T03:18:34.926Z iteration 0 passed in 187 ms

{
  id: '136091264370',
  clientOrderId: 'api',
  timestamp: 1646882314959,
  datetime: '2022-03-10T03:18:34.959Z',
  lastTradeTimestamp: undefined,
  status: 'open',
  symbol: 'XRP/USDT:USDT',
  type: 'limit',
  timeInForce: 'PO',
  postOnly: true,
  side: 'sell',
  price: 0.758,
  stopPrice: undefined,
  average: undefined,
  amount: 1,
  cost: 0,
  filled: 0,
  remaining: 1,
  fee: undefined,
  fees: [],
  trades: [],
  info: {
    id: '136091264370',
    contract: 'XRP_USDT',
    mkfr: '0',
    tkfr: '0.0005',
    tif: 'poc',
    is_reduce_only: false,
    create_time: '1646882314.959',
    price: '0.758',
    size: '-1',
    refr: '0.2',
    left: '-1',
    text: 'api',
    fill_price: '0',
    user: '6693577',
    status: 'open',
    is_liq: false,
    refu: '2436035',
    is_close: false,
    iceberg: '0'
  }
}
```

```
% gateio createOrder XRP/USDT:USDT limit sell 1 0.758 '{"timeInForce": "PO"}' --verbose
gateio.createOrder (XRP/USDT:USDT, limit, sell, 1, 0.758, [object Object])
fetch Request:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 
...
 {"contract":"XRP_USDT","size":-1,"price":"0.758","tif":"poc"} 

handleRestResponse:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 201 Created 
ResponseBody:
 {"id":136091414610,"contract":"XRP_USDT","mkfr":"0","tkfr":"0.0005","tif":"poc","is_reduce_only":false,"create_time":1646882366.093,"price":"0.758","size":-1,"refr":"0.2","left":-1,"text":"api","fill_price":"0","user":6693577,"status":"open","is_liq":false,"refu":2436035,"is_close":false,"iceberg":0}
```

```
sam@Sams-Mac-mini ccxt % gateio createOrder XRP/USDT:USDT postOnly buy 1 0.757 '{"reduceOnly": true}' --verbose
gateio.createOrder (XRP/USDT:USDT, postOnly, buy, 1, 0.757, [object Object])
fetch Request:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 
RequestHeaders:
RequestBody:
 {"contract":"XRP_USDT","size":1,"price":"0.757","reduce_only":true,"tif":"poc"} 

handleRestResponse:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 201 Created 
ResponseBody:
 {"id":136091666171,"contract":"XRP_USDT","mkfr":"0","tkfr":"0.0005","tif":"poc","is_reduce_only":true,"create_time":1646882463.902,"price":"0.757","size":1,"refr":"0.2","left":1,"text":"api","fill_price":"0","user":6693577,"status":"open","is_liq":false,"refu":2436035,"is_close":false,"iceberg":0}
```

## Error Responses

```
% gateio createOrder XRP/USDT:USDT market buy 1 0.757 '{"postOnly": true}' --verbose
gateio.createOrder (XRP/USDT:USDT, market, buy, 1, 0.757, [object Object])
InvalidOrder gateio postOnly orders cannot have type market
---------------------------------------------------
[InvalidOrder] gateio postOnly orders cannot have type market

    at isPostOnly                 ../../ccxt/ccxt/js/base/Exchange.js:2234  throw new InvalidOrder (this.id + ' postOnly orders cannot have type ' + type); 
    at createOrder                ../../ccxt/ccxt/js/gateio.js:2278         [ type, postOnly, timeInForce, params ] = this.isPostOnly (type, timeInForce, u…
    at processTicksAndRejections  internal/process/task_queues.js:95                                                                                        
    at async main                 ../../ccxt/ccxt/examples/js/cli.js:258    const result = await exchange[methodName] (... args)                            

gateio postOnly orders cannot have type market
```

```
% gateio createOrder XRP/USDT:USDT limit buy 1 0.757 '{"postOnly": true, "timeInForce": "IOC"}' --verbose
gateio.createOrder (XRP/USDT:USDT, limit, buy, 1, 0.757, [object Object])
InvalidOrder gateio postOnly orders cannot have timeInForce equal to IOC
---------------------------------------------------
[InvalidOrder] gateio postOnly orders cannot have timeInForce equal to IOC

    at isPostOnly                 ../../ccxt/ccxt/js/base/Exchange.js:2232  throw new InvalidOrder (this.id + ' postOnly orders cannot have timeInForce equ…
    at createOrder                ../../ccxt/ccxt/js/gateio.js:2278         [ type, postOnly, timeInForce, params ] = this.isPostOnly (type, timeInForce, u…
    at processTicksAndRejections  internal/process/task_queues.js:95                                                                                        
    at async main                 ../../ccxt/ccxt/examples/js/cli.js:258    const result = await exchange[methodName] (... args)                            

gateio postOnly orders cannot have timeInForce equal to IOC
Waiting for the debugger to disconnect...
```

```
sam@Sams-Mac-mini ccxt % gateio createOrder XRP/USDT:USDT postOnly buy 1 0.758 '{"reduceOnly": true}' --verbose
gateio.createOrder (XRP/USDT:USDT, postOnly, buy, 1, 0.758, [object Object])
fetch Request:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 
RequestHeaders:
 {
  'X-Gate-Channel-Id': 'ccxt',
  KEY: '',
  Timestamp: '1646882413',
  SIGN: '',
  'Content-Type': 'application/json'
} 
RequestBody:
 {"contract":"XRP_USDT","size":1,"price":"0.758","reduce_only":true,"tif":"poc"} 

handleRestResponse:
 gateio POST https://api.gateio.ws/api/v4/futures/usdt/orders 400 Bad Request 
ResponseHeaders:
 {
  Connection: 'keep-alive',
  'Content-Type': 'application/json',
  Date: 'Thu, 10 Mar 2022 03:20:13 GMT',
  Server: 'openresty',
  'Transfer-Encoding': 'chunked'
} 
ResponseBody:
 {"label":"ORDER_POC_IMMEDIATE","detail":"order price 0.758 while counter price 0.7575"}
 

InvalidOrder gateio {"label":"ORDER_POC_IMMEDIATE","detail":"order price 0.758 while counter price 0.7575"}

---------------------------------------------------
[InvalidOrder] gateio {"label":"ORDER_POC_IMMEDIATE","detail":"order price 0.758 while counter price 0.7575"}

    at throwExactlyMatchedException  ../../ccxt/ccxt/js/base/Exchange.js:635        throw new exact[string] (message)                                               
    at handleErrors                  ../../ccxt/ccxt/js/gateio.js:3549              this.throwExactlyMatchedException (this.exceptions['exact'], label, feedback);  
    at                               ../../ccxt/ccxt/js/base/Exchange.js:712        const skipFurtherErrorHandling = this.handleErrors (response.status, response.s…
    at processTicksAndRejections     internal/process/task_queues.js:95                                                                                             
    at async timeout                 ../../ccxt/ccxt/js/base/functions/time.js:203  return await Promise.race ([promise, expires.then (() => { throw new TimedOut (…
    at createOrder                   ../../ccxt/ccxt/js/gateio.js:2430              const response = await this[method] (this.deepExtend (request, params));        
    at async main                    ../../ccxt/ccxt/examples/js/cli.js:258         const result = await exchange[methodName] (... args)                            

gateio {"label":"ORDER_POC_IMMEDIATE","detail":"order price 0.758 while counter price 0.7575"}
```

# SPOT

```
% gateio createOrder XRP/USDT limit buy 2 0.755 '{"postOnly": true}' --verbose 
gateio.createOrder (XRP/USDT, limit, buy, 2, 0.755, [object Object])
RequestBody:
 {"currency_pair":"XRP_USDT","type":"limit","account":"spot","side":"buy","amount":"2","price":"0.755","time_in_force":"poc"} 

ResponseBody:
 {"id":"131322281764","text":"apiv4","create_time":"1646883760","update_time":"1646883760","create_time_ms":1646883760373,"update_time_ms":1646883760373,"status":"open","currency_pair":"XRP_USDT","type":"limit","account":"spot","side":"buy","amount":"2","price":"0.755","time_in_force":"poc","iceberg":"0","left":"2","fill_price":"0","filled_total":"0","fee":"0","fee_currency":"XRP","point_fee":"0","gt_fee":"0","gt_discount":false,"rebated_fee":"0","rebated_fee_currency":"USDT"} 

2022-03-10T03:42:40.392Z iteration 0 passed in 199 ms

{
  id: '131322281764',
  clientOrderId: 'apiv4',
  timestamp: 1646883760000,
  datetime: '2022-03-10T03:42:40.000Z',
  lastTradeTimestamp: 1646883760000,
  status: 'open',
  symbol: 'XRP/USDT',
  type: 'limit',
  timeInForce: 'PO',
  postOnly: true,
  side: 'buy',
  price: 0.755,
  stopPrice: undefined,
  average: undefined,
  amount: 2,
  cost: 0,
  filled: 0,
  remaining: 2,
  fee: undefined,
  fees: [
    { currency: 'GT', cost: '0' },
    { currency: 'XRP', cost: '0' },
    { currency: 'USDT', cost: '0' }
  ],
  trades: [],
  info: {
    id: '131322281764',
    text: 'apiv4',
    create_time: '1646883760',
    update_time: '1646883760',
    create_time_ms: '1646883760373',
    update_time_ms: '1646883760373',
    status: 'open',
    currency_pair: 'XRP_USDT',
    type: 'limit',
    account: 'spot',
    side: 'buy',
    amount: '2',
    price: '0.755',
    time_in_force: 'poc',
    iceberg: '0',
    left: '2',
    fill_price: '0',
    filled_total: '0',
    fee: '0',
    fee_currency: 'XRP',
    point_fee: '0',
    gt_fee: '0',
    gt_discount: false,
    rebated_fee: '0',
    rebated_fee_currency: 'USDT'
  }
}
```

```
% gateio createOrder XRP/USDT limit buy 2 0.755 '{"timeInForce": "PO"}' --verbose
gateio.createOrder (XRP/USDT, limit, buy, 2, 0.755, [object Object])
RequestBody:
 {"currency_pair":"XRP_USDT","type":"limit","account":"spot","side":"buy","amount":"2","price":"0.755","time_in_force":"poc"} 
 
ResponseBody:
 {"id":"131322677367","text":"apiv4","create_time":"1646883879","update_time":"1646883879","create_time_ms":1646883879380,"update_time_ms":1646883879380,"status":"open","currency_pair":"XRP_USDT","type":"limit","account":"spot","side":"buy","amount":"2","price":"0.755","time_in_force":"poc","iceberg":"0","left":"2","fill_price":"0","filled_total":"0","fee":"0","fee_currency":"XRP","point_fee":"0","gt_fee":"0","gt_discount":false,"rebated_fee":"0","rebated_fee_currency":"USDT"}
```

```
% gateio createPostOnlyOrder XRP/USDT buy 2 0.755 --verbose 
gateio.createPostOnlyOrder (XRP/USDT, buy, 2, 0.755)
RequestBody:
 {"currency_pair":"XRP_USDT","type":"limit","account":"spot","side":"buy","amount":"2","price":"0.755","time_in_force":"poc"} 
ResponseBody:
 {"id":"131323112897","text":"apiv4","create_time":"1646884023","update_time":"1646884023","create_time_ms":1646884023770,"update_time_ms":1646884023770,"status":"open","currency_pair":"XRP_USDT","type":"limit","account":"spot","side":"buy","amount":"2","price":"0.755","time_in_force":"poc","iceberg":"0","left":"2","fill_price":"0","filled_total":"0","fee":"0","fee_currency":"XRP","point_fee":"0","gt_fee":"0","gt_discount":false,"rebated_fee":"0","rebated_fee_currency":"USDT"} 
```

### CODE REVIEW

**Comment by frosty00** [comment]
for `createPostOnly` order, you are better off putting it in the params since that is supported by most of the exchanges, the `postOnly` limit arg is supported by very few

**Comment by frosty00** [comment]
so i'd suggest to edit it like this

```
createPostOnlyOrder
   return this.createOrder (...args, { postOnly: true })

**Comment by samgermain** [comment]
> for `createPostOnly` order, you are better off putting it in the params since that is supported by most of the exchanges, the `postOnly` limit arg is supported by very few
> 
> at least idk how far you have gotten with the `this.isPostOnly` unification so far

I could do that, it doesn't really matter with the way I've unified it. You can see the other exchanges I've done it for on [this PR](https://github.com/ccxt/ccxt/pull/12228)

**Comment by samgermain** [comment]
@frosty00 @kroitor Do we want to have the order type hard set to `limit`? Or should we leave it as a parameter?

File: php/async/Exchange.php [hunk 1]
```diff
@@ -358,10 +358,12 @@ public function sleep($milliseconds) {
...
 
-    public function create_post_only_order($symbol, $side, $amount, $price = null, $params = array()) {
+    public function create_post_only_order($symbol, $type, $side, $amount, $price, $params = array()) {
         if (!$this->has['createPostOnlyOrder']) {
             throw new NotSupported($this->id + 'create_post_only_order() is not supported yet');
         }
-        return yield $this.create_order($symbol, 'postOnly', $side, $amount, $price, $params);
+        $array = array('postOnly' => true);
+        $query = $this->extend($params, $array);
+        return yield $this.create_order($symbol, $type, $side, $amount, $price, $params);
```
Comment (frosty00) [inline php/async/Exchange.php:367]: this looks like a bug to me, `$this.create_order`

**Comment by samgermain** [comment]
@kroitor Are there any more changes you want with this PR?

**Comment by kroitor** [comment]
@samgermain i'll test and merge shortly, thx!

File: js/base/Exchange.js [hunk 2]
```diff
@@ -2222,4 +2224,42 @@ module.exports = class Exchange {
...
+     * @param {dict} params: Exchange specific params
+     * @returns {boolean}: true if a post only order, false otherwise
+     */
+    isPostOnly (type, timeInForce, exchangeSpecificOption, params = {}) {
+        let postOnly = this.safeValue2 (params, 'postOnly', 'post_only', false);
+        params = this.omit (params, ['post_only', 'postOnly']);
+        const timeInForceUpper = timeInForce.toUpperCase ();
+        const typeLower = type.toLowerCase ();
+        const ioc = timeInForceUpper === 'IOC';
+        const timeInForcePostOnly = timeInForceUpper === 'PO';
```
Comment (frosty00) [inline js/base/Exchange.js:2241]: I think the only allowable `timeInForce` for postOnly orders should be `PO`

Comment (samgermain) [reply js/base/Exchange.js:2241]: GTC is valid, there's also some exchanges that specify GTX, postOnly ...

Comment (samgermain) [reply js/base/Exchange.js:2242]: I added FOK, but GTC should stay

**Pull request merged**